core/gather/gatherers/trace-elements.js

```diff
@@ -65,6 +65,61 @@ class TraceElements extends BaseGatherer {
     if (name) this.animationIdToName.set(id, name);
   }
 
+  /**
+   * @param {LH.Artifacts.TraceEngineResult} traceEngineResult
+   * @param {string} mainFrameId
+   * @return {Promise<Array<{nodeId: number}>>}
+   */
+  static async getTraceEngineElements(traceEngineResult, mainFrameId) {
+    // Can only resolve elements for the latest insight set, which should correspond
+    // to the current frame id. Can't resolve elements for pages that are gone.
+    const insightSet = [...traceEngineResult.insights.values()]
+      .reverse().find(insightSet => insightSet.frameId === mainFrameId);
+    if (!insightSet) {
+      return [];
+    }
+
+    /**
+     * Execute `cb(obj, key)` on every object property (non-objects only), recursively.
+     * @param {any} obj
+     * @param {(obj: Record<string, string>, key: string) => void} cb
+     * @param {Set<object>} seen
+     */
+    function recursiveObjectEnumerate(obj, cb, seen) {
+      if (seen.has(seen)) {
+        return;
+      }
+
+      seen.add(obj);
+
+      if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
+        Object.keys(obj).forEach(key => {
+          if (typeof obj[key] === 'object') {
+            recursiveObjectEnumerate(obj[key], cb, seen);
+          } else {
+            cb(obj, key);
+          }
+        });
+      } else if (Array.isArray(obj)) {
+        obj.forEach(item => {
+          if (typeof item === 'object' || Array.isArray(item)) {
+            recursiveObjectEnumerate(item, cb, seen);
+          }
+        });
+      }
+    }
+
+    /** @type {number[]} */
+    const nodeIds = [];
+    recursiveObjectEnumerate(insightSet.model, (obj, key) => {
+      if (typeof obj[key] === 'number' && (key === 'nodeId' || key === 'node_id')) {
```

Comment:
nit: create a `NODE_ID_KEYS` constant in case there are more we need to add to this list